### DISCUSSION TITLE
Adding cell context. (like pkg mode or other function style contexts)

### DISCUSSION THREAD

**Discussion by rapus95** [fonsp/Pluto.jl#944] [Pluto development]
It'd be awesome if individual cells could be executed within a certain (user-defined) context.

As a very basic example there's the RCall macro `R"some arbitrary r code"` which executes R code. And there's also the package mode macro `pkg"add x"`. If I now could select to have a cell in R context, that is, automatically wrapped in R_str macro, that would be very cool. Other situations that come up to my mind could be a filehandler etc.

Idea on how to integrate user-defined context types:
Have 2 types of context:
1) function (this one just accepts the cell it is used as context for as a function)
2) macro (this one just accepts the cell by wrapping the code in a macro call)
3) stringmacro (this one is a special case of 2 which wraps everything into triple quotation marks before feeding into 2)

In order to have it still be ordinary julia code files, that context object follows the code pattern:
macro context:
`@macro(cell content)`

function context:
```julia
function() do args...
cell content
end
```

This is a very basic and first draw on such context mode and I'd really love if it'd at least be considered :D

**Comment by fonsp** [comment]
I am not sure what you mean. You can already use macros, function contexts, etc in Pluto cells:
```
R"""
x <- 123
"""
```

which is very cool! Can you edit your suggestion to be specific about how it differs from this existing functionality? Are you talking about a new GUI element? What do you mean with "automatically"?

**Reply by rapus95** [reply]
it's intended to be some visual flavor that hides the macro call from the actual textfield but instead shows the "context" as a colored cell background (which could mimic the repl mode color) and/or as a label which is placed somewhere relative to each cell. No Context which obviously is the default, shows everything as it is right now. But assigning a context to a cell, like pkg mode or R mode etc, will show that context label and an individual background and will make the cell content properly wrapped in code in the backend but unwrapped in the front-end.
